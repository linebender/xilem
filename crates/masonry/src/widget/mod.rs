// This software is licensed under Apache License 2.0 and distributed on an
// "as-is" basis without warranties of any kind. See the LICENSE file for
// details.

//! Common widgets.

#[allow(clippy::module_inception)]
mod widget;
mod widget_mut;
mod widget_pod;
mod widget_ref;
mod widget_state;

#[cfg(test)]
mod tests;

mod align;
mod button;
mod checkbox;
mod flex;
mod image;
mod label;
mod portal;
mod scroll_bar;
mod sized_box;
mod spinner;
mod split;

pub use self::image::Image;
pub use align::Align;
pub use button::Button;
pub use checkbox::Checkbox;
pub use flex::{Axis, CrossAxisAlignment, Flex, FlexParams, MainAxisAlignment};
pub use label::{Label, LineBreaking};
pub use portal::Portal;
pub use scroll_bar::ScrollBar;
pub use sized_box::SizedBox;
pub use spinner::Spinner;
pub use split::Split;
pub use widget_mut::WidgetMut;
pub use widget_pod::WidgetPod;
pub use widget_ref::WidgetRef;
pub use widget_state::WidgetState;

pub use sized_box::BackgroundBrush;
pub use widget::StoreInWidgetMut;
#[doc(hidden)]
pub use widget::{Widget, WidgetId};
use winit::window::CursorIcon;

/// Methods by which a widget can attempt to change focus state.
#[derive(Debug, Clone, Copy)]
pub(crate) enum FocusChange {
    /// The focused widget is giving up focus.
    Resign,
    /// A specific widget wants focus
    Focus(WidgetId),
    /// Focus should pass to the next focusable widget
    Next,
    /// Focus should pass to the previous focusable widget
    Previous,
}

/// The possible cursor states for a widget.
#[derive(Clone, Debug)]
pub(crate) enum CursorChange {
    /// No cursor has been set.
    Default,
    /// Someone set a cursor, but if a child widget also set their cursor then we'll use theirs
    /// instead of ours.
    Set(CursorIcon),
    /// Someone set a cursor, and we'll use it regardless of what the children say.
    Override(CursorIcon),
}

use crate::{Affine, Size};

// These are based on https://api.flutter.dev/flutter/painting/BoxFit-class.html
/// Strategies for inscribing a rectangle inside another rectangle.
#[derive(Clone, Copy, Default, PartialEq)]
pub enum FillStrat {
    /// As large as posible without changing aspect ratio of image and all of image shown
    #[default]
    Contain,
    /// As large as posible with no dead space so that some of the image may be clipped
    Cover,
    /// Fill the widget with no dead space, aspect ratio of widget is used
    Fill,
    /// Fill the hight with the images aspect ratio, some of the image may be clipped
    FitHeight,
    /// Fill the width with the images aspect ratio, some of the image may be clipped
    FitWidth,
    /// Do not scale
    None,
    /// Scale down to fit but do not scale up
    ScaleDown,
}

// TODO
impl CursorChange {
    pub fn cursor(&self) -> Option<CursorIcon> {
        match self {
            CursorChange::Set(c) | CursorChange::Override(c) => Some(*c),
            CursorChange::Default => None,
        }
    }
}

// TODO - Need to write tests for this, in a way that's relatively easy to visualize.

impl FillStrat {
    /// Calculate an origin and scale for an image with a given `FillStrat`.
    ///
    /// This takes some properties of a widget and a fill strategy and returns an affine matrix
    /// used to position and scale the image in the widget.
    pub fn affine_to_fill(self, parent: Size, fit_box: Size) -> Affine {
        let raw_scalex = parent.width / fit_box.width;
        let raw_scaley = parent.height / fit_box.height;

        let (scalex, scaley) = match self {
            FillStrat::Contain => {
                let scale = raw_scalex.min(raw_scaley);
                (scale, scale)
            }
            FillStrat::Cover => {
                let scale = raw_scalex.max(raw_scaley);
                (scale, scale)
            }
            FillStrat::Fill => (raw_scalex, raw_scaley),
            FillStrat::FitHeight => (raw_scaley, raw_scaley),
            FillStrat::FitWidth => (raw_scalex, raw_scalex),
            FillStrat::ScaleDown => {
                let scale = raw_scalex.min(raw_scaley).min(1.0);
                (scale, scale)
            }
            FillStrat::None => (1.0, 1.0),
        };

        let origin_x = (parent.width - (fit_box.width * scalex)) / 2.0;
        let origin_y = (parent.height - (fit_box.height * scaley)) / 2.0;

        Affine::new([scalex, 0., 0., scaley, origin_x, origin_y])
    }
}

// TODO - remove prelude
#[allow(missing_docs)]
pub mod prelude {
    #[doc(hidden)]
    pub use crate::{
        BoxConstraints, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, PointerEvent, Size,
        StatusChange, TextEvent, Widget, WidgetId,
    };
}
