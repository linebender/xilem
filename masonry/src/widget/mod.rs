// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Common widgets.

#[allow(clippy::module_inception)]
pub(crate) mod widget;
mod widget_mut;
mod widget_pod;
mod widget_ref;
mod widget_state;

#[cfg(test)]
mod tests;

mod align;
mod button;
mod checkbox;
mod flex;
mod grid;
mod image;
mod label;
mod portal;
mod progress_bar;
mod prose;
mod root_widget;
mod scroll_bar;
mod sized_box;
mod spinner;
mod split;
mod textbox;
mod variable_label;
mod widget_arena;
mod taffy_layout;

pub use self::image::Image;
pub use align::Align;
pub use button::Button;
pub use checkbox::Checkbox;
pub use flex::{CrossAxisAlignment, Flex, FlexParams, MainAxisAlignment};
pub use grid::{Grid, GridParams};
pub use taffy_layout::TaffyLayout;
pub use label::{Label, LineBreaking};
pub use portal::Portal;
pub use progress_bar::ProgressBar;
pub use prose::Prose;
pub use root_widget::RootWidget;
pub use scroll_bar::ScrollBar;
pub use sized_box::SizedBox;
pub use spinner::Spinner;
pub use split::Split;
pub use textbox::Textbox;
pub use variable_label::VariableLabel;
pub use widget_mut::WidgetMut;
pub use widget_pod::WidgetPod;
pub use widget_ref::WidgetRef;
pub use widget::Axis;

pub(crate) use widget_arena::WidgetArena;
pub(crate) use widget_state::WidgetState;

use crate::{Affine, Size};

// These are based on https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit
/// Strategies for inscribing a rectangle inside another rectangle.
#[derive(Clone, Copy, Default, PartialEq)]
pub enum ObjectFit {
    /// As large as possible without changing aspect ratio of image and all of image shown
    #[default]
    Contain,
    /// As large as possible with no dead space so that some of the image may be clipped
    Cover,
    /// Fill the widget with no dead space, aspect ratio of widget is used
    Fill,
    /// Fill the height with the images aspect ratio, some of the image may be clipped
    FitHeight,
    /// Fill the width with the images aspect ratio, some of the image may be clipped
    FitWidth,
    /// Do not scale
    None,
    /// Scale down to fit but do not scale up
    ScaleDown,
}

// TODO - Need to write tests for this, in a way that's relatively easy to visualize.

impl ObjectFit {
    /// Calculate an origin and scale for an image with a given `ObjectFit`.
    ///
    /// This takes some properties of a widget and an object fit and returns an affine matrix
    /// used to position and scale the image in the widget.
    pub fn affine_to_fill(self, parent: Size, fit_box: Size) -> Affine {
        let raw_scalex = parent.width / fit_box.width;
        let raw_scaley = parent.height / fit_box.height;

        let (scalex, scaley) = match self {
            ObjectFit::Contain => {
                let scale = raw_scalex.min(raw_scaley);
                (scale, scale)
            }
            ObjectFit::Cover => {
                let scale = raw_scalex.max(raw_scaley);
                (scale, scale)
            }
            ObjectFit::Fill => (raw_scalex, raw_scaley),
            ObjectFit::FitHeight => (raw_scaley, raw_scaley),
            ObjectFit::FitWidth => (raw_scalex, raw_scalex),
            ObjectFit::ScaleDown => {
                let scale = raw_scalex.min(raw_scaley).min(1.0);
                (scale, scale)
            }
            ObjectFit::None => (1.0, 1.0),
        };

        let origin_x = (parent.width - (fit_box.width * scalex)) / 2.0;
        let origin_y = (parent.height - (fit_box.height * scaley)) / 2.0;

        Affine::new([scalex, 0., 0., scaley, origin_x, origin_y])
    }
}
