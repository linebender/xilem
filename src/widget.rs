// Copyright 2022 The Druid Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

pub mod align;
pub mod button;
mod contexts;
mod core;
//pub mod layout_observer;
//pub mod list;
pub mod piet_scene_helpers;
mod raw_event;
//pub mod scroll_view;
pub mod compose_style;
pub mod text;
//pub mod vstack;

use std::any::Any;
use std::ops::{Deref, DerefMut};

use glazier::kurbo::{Rect, Size};
use piet_scene::SceneBuilder;

use self::contexts::LifeCycleCx;
pub use self::contexts::{AlignCx, CxState, EventCx, LayoutCx, PaintCx, PreparePaintCx, UpdateCx};
pub use self::core::Pod;
pub(crate) use self::core::{PodFlags, WidgetState};
pub use self::raw_event::{LifeCycle, RawEvent};

use self::align::SingleAlignment;

/// A basic widget trait.
pub trait Widget {
    fn event(&mut self, cx: &mut EventCx, event: &RawEvent);

    /// Propagate a lifecycle event.
    ///
    /// I am not convinced this needs to be distinct from `event`. For the
    /// moment, we're following existing Druid.
    fn lifecycle(&mut self, cx: &mut LifeCycleCx, event: &LifeCycle);

    fn update(&mut self, cx: &mut UpdateCx);

    /// Compute intrinsic sizes.
    ///
    /// This method will be called once on widget creation and then on
    /// REQUEST_UPDATE.
    fn measure(&mut self, cx: &mut LayoutCx) -> (Size, Size);

    /// Compute size given proposed size.
    ///
    /// The value will be memoized given the proposed size, invalidated
    /// on REQUEST_UPDATE. It can count on prelayout being completed.
    fn layout(&mut self, cx: &mut LayoutCx, proposed_size: Size) -> Size;

    /// Query for an alignment.
    ///
    /// This method can count on layout already having been completed.
    #[allow(unused)]
    fn align(&self, cx: &mut AlignCx, alignment: SingleAlignment) {}

    /// Prepare for painting.
    ///
    /// This method is currently a bit of a hack. It's similar to the one in
    /// Druid, which is for incremental repaint, but the primary purpose of
    /// this one is virtualized scrolling.
    ///
    /// The fact that `cx` is LayoutCx is just laziness, it should have its
    /// own cx. And the main methods on that cx should be for region-based
    /// invalidation.
    #[allow(unused)]
    fn prepare_paint(&mut self, cx: &mut LayoutCx, visible: Rect) {}

    fn paint(&mut self, cx: &mut PaintCx, builder: &mut SceneBuilder);
}

pub trait AnyWidget: Widget {
    fn as_any(&self) -> &dyn Any;

    fn as_any_mut(&mut self) -> &mut dyn Any;

    fn type_name(&self) -> &'static str;
}

impl<W: Widget + 'static> AnyWidget for W {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn type_name(&self) -> &'static str {
        std::any::type_name::<Self>()
    }
}

impl Widget for Box<dyn AnyWidget> {
    fn event(&mut self, cx: &mut EventCx, event: &RawEvent) {
        self.deref_mut().event(cx, event);
    }

    fn lifecycle(&mut self, cx: &mut LifeCycleCx, event: &LifeCycle) {
        self.deref_mut().lifecycle(cx, event);
    }

    fn update(&mut self, cx: &mut UpdateCx) {
        self.deref_mut().update(cx);
    }

    fn measure(&mut self, cx: &mut LayoutCx) -> (Size, Size) {
        self.deref_mut().measure(cx)
    }

    fn layout(&mut self, cx: &mut LayoutCx, proposed_size: Size) -> Size {
        self.deref_mut().layout(cx, proposed_size)
    }

    fn align(&self, cx: &mut AlignCx, alignment: SingleAlignment) {
        self.deref().align(cx, alignment);
    }

    fn prepare_paint(&mut self, cx: &mut PreparePaintCx, visible: Rect) {
        self.deref_mut().prepare_paint(cx, visible)
    }

    fn paint(&mut self, cx: &mut PaintCx, builder: &mut SceneBuilder) {
        self.deref_mut().paint(cx, builder);
    }
}

pub trait WidgetTuple {
    fn length(&self) -> usize;

    // Follows Panoramix; rethink to reduce allocation
    // Maybe SmallVec?
    fn widgets_mut(&mut self) -> Vec<&mut dyn AnyWidget>;
}

macro_rules! impl_widget_tuple {
    ( $n: tt; $( $WidgetType:ident),* ; $( $index:tt ),* ) => {
        impl< $( $WidgetType: AnyWidget ),* > WidgetTuple for ( $( $WidgetType, )* ) {
            fn length(&self) -> usize {
                $n
            }

            fn widgets_mut(&mut self) -> Vec<&mut dyn AnyWidget> {
                let mut v: Vec<&mut dyn AnyWidget> = Vec::with_capacity(self.length());
                $(
                v.push(&mut self.$index);
                )*
                v
            }

        }
    }
}

impl_widget_tuple!(1; W0; 0);
impl_widget_tuple!(2; W0, W1; 0, 1);
impl_widget_tuple!(3; W0, W1, W2; 0, 1, 2);
impl_widget_tuple!(4; W0, W1, W2, W3; 0, 1, 2, 3);
impl_widget_tuple!(5; W0, W1, W2, W3, W4; 0, 1, 2, 3, 4);
impl_widget_tuple!(6; W0, W1, W2, W3, W4, W5; 0, 1, 2, 3, 4, 5);
impl_widget_tuple!(7; W0, W1, W2, W3, W4, W5, W6; 0, 1, 2, 3, 4, 5, 6);
impl_widget_tuple!(8;
    W0, W1, W2, W3, W4, W5, W6, W7;
    0, 1, 2, 3, 4, 5, 6, 7
);
