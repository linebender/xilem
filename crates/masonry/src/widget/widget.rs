// This software is licensed under Apache License 2.0 and distributed on an
// "as-is" basis without warranties of any kind. See the LICENSE file for
// details.

use std::any::Any;
use std::num::NonZeroU64;
use std::ops::{Deref, DerefMut};
use std::sync::atomic::{AtomicU64, Ordering};

use smallvec::SmallVec;
use tracing::{trace_span, Span};
use vello::Scene;

use crate::event::StatusChange;
use crate::event::{PointerEvent, TextEvent};
use crate::widget::WidgetRef;
use crate::{
    AsAny, BoxConstraints, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, Point, Size,
};

/// A unique identifier for a single [`Widget`].
///
/// `WidgetId`s are generated automatically for all widgets in the widget tree.
/// More specifically, each [`WidgetPod`](crate::WidgetPod) has a unique `WidgetId`.
///
/// These ids are used internally to route events, and can be used to communicate
/// between widgets, by submitting a command (as with [`EventCtx::submit_command`])
/// and passing a `WidgetId` as the [`Target`](crate::Target).
///
/// A widget can retrieve its id via methods on the various contexts, such as
/// [`LifeCycleCtx::widget_id`].
///
/// ## Explicit `WidgetId`s.
///
/// Sometimes, you may want to construct a widget, in a way that lets you know its id,
/// so you can refer to the widget later. You can use [`WidgetPod::new_with_id`](crate::WidgetPod::new_with_id) to pass
/// an id to the WidgetPod you're creating; various widgets which have methods to create
/// children may have variants taking ids as parameters.
///
/// If you set a `WidgetId` directly, you are resposible for ensuring that it
/// is unique. Two widgets must not be created with the same id.
#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
pub struct WidgetId(NonZeroU64);

// TODO - Add tutorial: implementing a widget - See issue #5
/// The trait implemented by all widgets.
///
/// For details on how to implement this trait, see tutorial **(TODO)**
///
/// Whenever external events affect the given widget, methods `[on_event]`,
/// `[on_status_change]` and `[on_lifecycle]` are called. Later on, when the
/// widget is laid out and displayed, methods `[layout]` and `[paint]` are called.
///
/// These trait methods are provided with a corresponding context. The widget can
/// request things and cause actions by calling methods on that context.
///
/// Widgets also have a `children()` method. Leaf widgets return an empty array,
/// whereas container widgets return an array of [`WidgetRef`]. Container widgets
/// have some validity invariants to maintain regarding their children. See TUTORIAL_2
/// for details **(TODO)**.
///
/// Generally speaking, widgets aren't used directly. They are stored in
/// [`WidgetPods`](crate::WidgetPod). Widget methods are called by WidgetPods, and the
/// widget is mutated either during a method call (eg `on_event` or `lifecycle`) or
/// through a [`WidgetMut`](crate::widget::WidgetMut). See tutorials for detail.
pub trait Widget: AsAny {
    /// Handle an event - usually user interaction.
    ///
    /// A number of different events (in the [`Event`] enum) are handled in this
    /// method call. A widget can handle these events in a number of ways:
    /// requesting things from the [`EventCtx`], mutating the data, or submitting
    /// a [`Command`](crate::Command).
    fn on_pointer_event(&mut self, ctx: &mut EventCtx, event: &PointerEvent);
    fn on_text_event(&mut self, ctx: &mut EventCtx, event: &TextEvent);

    #[allow(missing_docs)]
    fn on_status_change(&mut self, ctx: &mut LifeCycleCtx, event: &StatusChange);

    /// Handle a lifecycle notification.
    ///
    /// This method is called to notify your widget of certain special events,
    /// (available in the [`LifeCycle`] enum) that are generally related to
    /// changes in the widget graph or in the state of your specific widget.
    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle);

    /// Compute layout.
    ///
    /// A leaf widget should determine its size (subject to the provided
    /// constraints) and return it.
    ///
    /// A container widget will recursively call [`WidgetPod::layout`](crate::WidgetPod::layout) on its
    /// child widgets, providing each of them an appropriate box constraint,
    /// compute layout, then call [`LayoutCtx::place_child`] on each of its children.
    /// Finally, it should return the size of the container. The container
    /// can recurse in any order, which can be helpful to, for example, compute
    /// the size of non-flex widgets first, to determine the amount of space
    /// available for the flex widgets.
    ///
    /// For efficiency, a container should only invoke layout of a child widget
    /// once, though there is nothing enforcing this.
    ///
    /// The layout strategy is strongly inspired by Flutter.
    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size;

    /// Paint the widget appearance.
    ///
    /// Container widgets can paint a background before recursing to their
    /// children, or annotations (for example, scrollbars) by painting
    /// afterwards. In addition, they can apply masks and transforms on
    /// the render context, which is especially useful for scrolling.
    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene);

    /// Return references to this widget's children.
    ///
    /// Leaf widgets return an empty array. Container widgets return references to
    /// their children.
    ///
    /// This methods has some validity invariants. A widget's children list must be
    /// consistent. If children are added or removed, the parent widget should call
    /// `children_changed` on one of the Ctx parameters. Container widgets are also
    /// responsible for calling the main methods (on_event, lifecycle, layout, paint)
    /// on their children.
    fn children(&self) -> SmallVec<[WidgetRef<'_, dyn Widget>; 16]>;

    /// Return a span for tracing.
    ///
    /// As methods recurse through the widget tree, trace spans are added for each child
    /// widget visited, and popped when control flow goes back to the parent. This method
    /// returns a static span (that you can use to filter traces and logs).
    fn make_trace_span(&self) -> Span {
        trace_span!("Widget", r#type = self.short_type_name())
    }

    /// Return a small string representing important info about this widget instance.
    ///
    /// When using [`WidgetRef`]'s [Debug](std::fmt::Debug) implementation, widgets
    /// will be displayed as a tree of values. Widgets which return a non-null value in
    /// `get_debug_text` will appear with that text next to their type name. This can
    /// be eg a label's text, or whether a checkbox is checked.
    fn get_debug_text(&self) -> Option<String> {
        None
    }

    // --- Auto-generated implementations ---

    /// Return which child, if any, has the given `pos` in its layout rect.
    ///
    /// The child return is a direct child, not eg a grand-child. The position is in
    /// relative cordinates. (Eg `(0,0)` is the top-left corner of `self`).
    ///
    /// Has a default implementation, that can be overriden to search children more
    /// efficiently.
    fn get_child_at_pos(&self, pos: Point) -> Option<WidgetRef<'_, dyn Widget>> {
        // layout_rect() is in parent coordinate space
        self.children()
            .into_iter()
            .find(|child| child.state().layout_rect().contains(pos))
    }

    /// Get the (verbose) type name of the widget for debugging purposes.
    /// You should not override this method.
    #[doc(hidden)]
    fn type_name(&self) -> &'static str {
        std::any::type_name::<Self>()
    }

    /// Get the (abridged) type name of the widget for debugging purposes.
    /// You should not override this method.
    #[doc(hidden)]
    fn short_type_name(&self) -> &'static str {
        let name = self.type_name();
        name.split('<')
            .next()
            .unwrap_or(name)
            .split("::")
            .last()
            .unwrap_or(name)
    }

    // FIXME
    /// Cast as Any.
    ///
    /// Mainly intended to be overriden in `Box<dyn Widget>`.
    #[doc(hidden)]
    fn as_any(&self) -> &dyn Any {
        self.as_dyn_any()
    }

    // FIXME
    /// Cast as Any.
    ///
    /// Mainly intended to be overriden in `Box<dyn Widget>`.
    #[doc(hidden)]
    fn as_mut_any(&mut self) -> &mut dyn Any {
        self.as_mut_dyn_any()
    }
}

/// Trait that widgets must implement to be in [`WidgetMut`](crate::widget::WidgetMut).
///
/// This trait should usually be implemented with [`declare_widget`](crate::declare_widget).
#[allow(missing_docs)]
pub trait StoreInWidgetMut: Widget {
    type Mut<'a>: Deref<Target = Self>;

    fn from_widget_and_ctx<'a>(
        widget: &'a mut Self,
        ctx: crate::contexts::WidgetCtx<'a>,
    ) -> Self::Mut<'a>;

    fn get_widget<'s: 'r, 'a: 'r, 'r>(widget_mut: &'s mut Self::Mut<'a>) -> &'r mut Self {
        Self::get_widget_and_ctx(widget_mut).0
    }

    fn get_ctx<'s: 'r, 'a: 'r, 'r>(
        widget_mut: &'s mut Self::Mut<'a>,
    ) -> &'r mut crate::contexts::WidgetCtx<'a> {
        Self::get_widget_and_ctx(widget_mut).1
    }

    fn get_widget_and_ctx<'s: 'r, 'a: 'r, 'r>(
        widget_mut: &'s mut Self::Mut<'a>,
    ) -> (&'r mut Self, &'r mut crate::contexts::WidgetCtx<'a>);
}

// TODO - Generate a struct instead. See #27.
/// Declare a mutable reference type for your widget.
///
/// The general syntax is:
///
/// ```ignore
/// declare_widget!(MyWidgetMut, MyWidget);
/// ```
///
/// where `MyWidget` is the name of your widget, and `MyWidgetMut` is an arbitrary
/// name (it can be exported for documentation purposes, but it's not going to be
/// instanced directly).
///
/// The above macro call will produce something like this:
///
/// ```ignore
/// pub struct MyWidgetMut<'a>(WidgetCtx<'a>, &'a mut MyWidget);
///
/// impl StoreInWidgetMut for MyWidget {
///     type Mut<'a> = MyWidgetMut<'a>;
/// }
/// ```
///
/// Because of `WidgetMut`'s [Deref] implementation, any methods implemented on
/// `MyWidgetMut` will thus be usable from `WidgetMut<MyWidget>`.
///
/// **Note:** This is all a huge hack to compensate for the lack of
/// [arbitrary self types](https://github.com/rust-lang/rust/issues/44874). What we
/// would really want is for that feature to be stable.
///
/// ## Generic widgets
///
/// If a widget type has generic arguments, the syntax becomes:
///
/// ```ignore
/// declare_widget!(FoobarMut, Foobar<A, B, C>);
/// ```
///
/// If these arguments have bounds, the syntax becomes:
///
/// ```ignore
/// declare_widget!(FoobarMut, Foobar<A: (SomeTrait), B: (SomeTrait + OtherTrait), C>);
/// ```
///
/// Yes, that is extremely annoying. Sorry about that.
#[macro_export]
macro_rules! declare_widget {
    ($WidgetNameMut:ident, $WidgetName:ident) => {
        $crate::declare_widget!($WidgetNameMut, $WidgetName<>);
    };

    ($WidgetNameMut:ident, $WidgetName:ident<$($Arg:ident $(: ($($Bound:tt)*))?),*>) => {
        pub struct $WidgetNameMut<'a, $($Arg $(: $($Bound)*)?),*>{
            ctx: $crate::WidgetCtx<'a>,
            widget: &'a mut $WidgetName<$($Arg),*>
        }

        impl<$($Arg $(: $($Bound)*)?),*> $crate::widget::StoreInWidgetMut for $WidgetName<$($Arg),*> {
            type Mut<'a> = $WidgetNameMut<'a, $($Arg),*>;

            fn get_widget_and_ctx<'s: 'r, 'a: 'r, 'r>(
                widget_mut: &'s mut Self::Mut<'a>,
            ) -> (&'r mut Self, &'r mut $crate::WidgetCtx<'a>) {
                (widget_mut.widget, &mut widget_mut.ctx)
            }

            fn from_widget_and_ctx<'a>(
                widget: &'a mut Self,
                ctx: $crate::WidgetCtx<'a>,
            ) -> Self::Mut<'a> {
                $WidgetNameMut { ctx, widget }
            }
        }

        impl<'a, $($Arg $(: $($Bound)*)?),*> ::std::ops::Deref for $WidgetNameMut<'a, $($Arg),*> {
            type Target = $WidgetName<$($Arg),*>;

            fn deref(&self) -> &Self::Target {
                self.widget
            }
        }
    };
}

#[cfg(not(tarpaulin_include))]
impl WidgetId {
    /// Allocate a new, unique `WidgetId`.
    ///
    /// All widgets are assigned ids automatically; you should only create
    /// an explicit id if you need to know it ahead of time, for instance
    /// if you want two sibling widgets to know each others' ids.
    ///
    /// You must ensure that a given `WidgetId` is only ever used for one
    /// widget at a time.
    pub fn next() -> WidgetId {
        static WIDGET_ID_COUNTER: AtomicU64 = AtomicU64::new(1);
        let id = WIDGET_ID_COUNTER.fetch_add(1, Ordering::Relaxed);
        WidgetId(id.try_into().unwrap())
    }

    // TODO - Remove
    /// Create a reserved `WidgetId`, suitable for reuse.
    ///
    /// The caller is responsible for ensuring that this ID is in fact assigned
    /// to a single widget at any time, or your code may become haunted.
    ///
    /// The actual inner representation of the returned `WidgetId` will not
    /// be the same as the raw value that is passed in; it will be
    /// `u64::max_value() - raw`.
    #[allow(unsafe_code)]
    pub const fn reserved(raw: u16) -> WidgetId {
        let id = u64::max_value() - raw as u64;
        // safety: by construction this can never be zero.
        WidgetId(unsafe { std::num::NonZeroU64::new_unchecked(id) })
    }

    pub(crate) fn to_raw(self) -> u64 {
        self.0.into()
    }
}

// TODO - remove
impl Widget for Box<dyn Widget> {
    fn on_pointer_event(&mut self, ctx: &mut EventCtx, event: &PointerEvent) {
        self.deref_mut().on_pointer_event(ctx, event)
    }

    fn on_text_event(&mut self, ctx: &mut EventCtx, event: &TextEvent) {
        self.deref_mut().on_text_event(ctx, event)
    }

    fn on_status_change(&mut self, ctx: &mut LifeCycleCtx, event: &StatusChange) {
        self.deref_mut().on_status_change(ctx, event)
    }

    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle) {
        self.deref_mut().lifecycle(ctx, event);
    }

    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
        self.deref_mut().layout(ctx, bc)
    }

    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
        self.deref_mut().paint(ctx, scene);
    }

    fn type_name(&self) -> &'static str {
        self.deref().type_name()
    }

    fn children(&self) -> SmallVec<[WidgetRef<'_, dyn Widget>; 16]> {
        self.deref().children()
    }

    fn make_trace_span(&self) -> Span {
        self.deref().make_trace_span()
    }

    fn get_debug_text(&self) -> Option<String> {
        self.deref().get_debug_text()
    }

    fn as_any(&self) -> &dyn Any {
        self.deref().as_dyn_any()
    }

    fn as_mut_any(&mut self) -> &mut dyn Any {
        self.deref_mut().as_mut_dyn_any()
    }
}

// We use alias type because macro doesn't accept braces except in some cases.
type BoxWidget = Box<dyn Widget>;
crate::declare_widget!(BoxWidgetMut, BoxWidget);
