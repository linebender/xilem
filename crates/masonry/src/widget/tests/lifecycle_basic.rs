// This software is licensed under Apache License 2.0 and distributed on an
// "as-is" basis without warranties of any kind. See the LICENSE file for
// details.

#![allow(unused_imports)]

use insta::assert_debug_snapshot;

use crate::testing::{
    widget_ids, Record, Recording, ReplaceChild, TestHarness, TestWidgetExt as _,
};
use crate::widget::{Flex, Label, SizedBox};
use crate::*;

#[test]
fn app_creation() {
    let record = Recording::default();
    let widget = SizedBox::empty().record(&record);

    let _harness = TestHarness::create(widget);

    let record = record.drain();
    assert_debug_snapshot!(record);
}

// FIXME - Need to figure out this test
#[ignore]
#[cfg(FALSE)]
/// Test that lifecycle events are sent correctly to a child added during event
/// handling
#[test]
fn adding_child() {
    let record = Recording::default();
    let record_new_child = Recording::default();
    let record_new_child2 = record_new_child.clone();

    let replacer = ReplaceChild::new(Label::new(""), move || {
        Flex::row()
            .with_child(Label::new("hello"))
            .with_child(Label::new("world"))
            .with_child(Label::new("hi"))
            .with_child(Label::new("").record(&record_new_child2))
    });

    let widget = Flex::row()
        .with_child(Label::new("hi").record(&record))
        .with_child(replacer);

    let mut harness = TestHarness::create(widget);
    record.clear();

    assert!(record_new_child.is_empty());

    harness.submit_command(REPLACE_CHILD);
    assert!(matches!(record.next(), Record::E(Event::Command(_))));

    let record_new_child = record_new_child.drain();
    assert_debug_snapshot!(record_new_child);
}

#[test]
fn child_tracking() {
    let [id_1, id_2, id_3, id_4] = widget_ids();

    let widget = SizedBox::new_with_id(
        SizedBox::new_with_id(
            Flex::row()
                .with_child_id(SizedBox::empty(), id_1)
                .with_child_id(SizedBox::empty(), id_2),
            id_3,
        ),
        id_4,
    );

    let harness = TestHarness::create(widget);

    let root_state = harness.get_widget(id_4).state();
    assert_eq!(root_state.children.entry_count(), 3);
    assert!(root_state.children.may_contain(&id_1));
    assert!(root_state.children.may_contain(&id_2));
    assert!(root_state.children.may_contain(&id_3));

    let child_state = harness.get_widget(id_3).state();
    assert!(child_state.children.may_contain(&id_1));
    assert!(child_state.children.may_contain(&id_2));
    assert_eq!(child_state.children.entry_count(), 2);
}

/// Test that all children are registered correctly after a child is replaced.
#[test]
#[cfg(FALSE)]
fn register_after_adding_child() {
    let [id_1, id_2, id_3, id_4, id_5, id_6, id_8] = widget_ids();

    let replacer = ReplaceChild::new(Label::new("hello").with_id(id_1), move || {
        SizedBox::new_with_id(
            Flex::row()
                .with_child_id(SizedBox::empty(), id_2)
                .with_child_id(SizedBox::empty(), id_3),
            id_4,
        )
    });

    let widget = Flex::row()
        .with_child_id(Label::new("hi"), id_8)
        .with_child_id(replacer, id_6)
        .with_id(id_5);

    let mut harness = TestHarness::create(widget);

    let root_state = harness.get_widget(id_5).state();
    assert!(root_state.children.may_contain(&id_6));
    assert!(root_state.children.may_contain(&id_1));
    assert!(root_state.children.may_contain(&id_8));

    harness.submit_command(REPLACE_CHILD);

    let root_state = harness.get_widget(id_5).state();
    assert!(root_state.children.may_contain(&id_6));
    assert!(root_state.children.may_contain(&id_8));
    assert!(root_state.children.may_contain(&id_4));
    assert!(root_state.children.may_contain(&id_2));
    assert!(root_state.children.may_contain(&id_3));
}
