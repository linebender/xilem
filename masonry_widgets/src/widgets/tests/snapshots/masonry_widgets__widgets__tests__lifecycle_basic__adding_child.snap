---
source: masonry/src/widget/tests/lifecycle_basic.rs
expression: record_new_child

---
[
    L(
        WidgetAdded,
    ),
    L(
        BuildFocusChain,
    ),
    Layout(
        4.0WÃ—18.0H,
    ),
    L(
        Internal(
            ParentWindowOrigin,
        ),
    ),
]
